BrowserPage Resturcturing

dos077 opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #1052 

## Description
<!-- Concisely describe what the pull request does. -->
Restructure BrowserPage view and its searchGridFilter component to fix some positioning bugs/improve reusability.

## Technical details
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
- AppModal:
  1.  remove close button, the close button is always inside the template content
  2. remove id prop, vue has its own dom management system, having an id that no one should call query on and may lead to duplicate warning is not a good idea
  3. added visible prop, controls whether the modal shows or not
  4. document listener when visible for keyup, so ESC will trigger $emit close
  5. changed the modal size to the original designs, otherwise it looks weird

- SearchGridFilter
  1. change top el from aside to div, the aside tag is not place next to main in browsePage, I think that makes more semantic sense.
  2. remove the duplicate conditional rendering, now the component just returns a FiltersList container.

- BrowsePage
  1. change the search column from div to aside
  2. added app-modal component
  3. use css class is-hidden-${breakpoint} to conditionally render between filters in modal or aside

A few more things I should point out about why so much changes:
1. Originally the filter modal (fix position) is nested inside a search column that has sticky position property. This is will not render correctly.
2. Bulma's applies { overflow-x: hidden; overflow-y: scroll } to most(all?)elements. This makes nesting sticky position element impossible, not that's easy to begin with.

## Tests
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete the section entirely. -->
All test passed, plus manually check all rendering in 1280px desktop/ipad horizontal/ipad vertical windows.

## Screenshots
<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the *default* branch of the repository (`main` or `master`).
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

zackkrida commented on src/components/AppModal.vue:
```diff
@@ -20,14 +17,33 @@
 export default {
   name: 'app-modal',
-  props: ['id'],
+  props: {
+    visible: Boolean,
+  },
+  watch: {
+    visible: {
+      handler(to) {
+        if (to) document.addEventListener('keyup', this.checkKey)
+        else document.removeEventListener('keyup', this.checkKey)
+      },
+      immediate: true,
+    },
+  },
+  destroyed() {
+    document.removeEventListener('keyup', this.checkKey)
+  },
+  methods: {
+    checkKey(e) {
+      if (e.keyCode === 27) this.$emit('close')
+    },
+  },
```
Modal visibility will always be toggled by `v-show` on the `<app-modal />`. For that reason I would prefer to add/remove the key listener on mount/destroy instead of maintaining an internal 'visible' state.

zackkrida commented on src/components/AppModal.vue:
```diff
@@ -20,14 +17,33 @@
 .modal {
-  max-width: 80vw;
-  max-height: 84vh;
+  width: 21.875rem;
+  max-height: 37rem;
```
This modal is a reusable component. The existing sizes are used here: https://github.com/creativecommons/cccatalog-frontend/blob/develop/src/components/MetaSearch/MetaSearchModal.vue

I would prefer if you added a `size` prop with `small/large` variants, and applied your new sizes on the 'small' version and the existing sizes on the 'large'.

zackkrida commented on src/pages/BrowsePage.vue:
```diff
@@ -2,14 +2,19 @@
-        :class="`column is-narrow grid-sidebar is-paddingless ${
-          filtersExpandedByDefault ? 'full-height-sticky' : ''
-        }`"
+      <div class="is-hidden-desktop">
+        <app-modal :visible="isFilterVisible" @close="onToggleSearchGridFilter">
+          <search-grid-filter @onSearchFilterChanged="onSearchFormSubmit" />
+        </app-modal>
+      </div>
+      <aside
         v-if="isFilterVisible"
```
This all looks great 👍

zackkrida commented on src/styles/results-page.scss:
```diff
@@ -21,6 +21,8 @@
   width: 21.875rem;
 
   @include touch {
+    position: relative;
     width: 0;
+    height: 0;
```
👍

dos077 commented:
My bad for not checking where else the AppModal is used. However, after looking over code, I think the inside card styling should be move out of the modal completely, since that will hinder re-usability by making design decision on the template content. Here are some examples of how other component libraries design their modal/dialog:

[Vuetify](https://vuetifyjs.com/en/components/dialogs/)
[RMWC](https://rmwc.io/dialogs)
[Quasar](https://quasar.dev/vue-components/dialog)

Most popular and widely adapted pattern of modal is an overlay with a centered container, so virtually anything can fit in there. I think the best practice is to move the size and color style and close button render into the template content itself, and let the AppModal be as styling agnostic as possible.

As to the visible prop, it's also a fairly standard practice, since it's the only variable the modal cares about. Only divide seems to whether the modal can change its visible state (in the prop implementation that's disallow, only parent can change visible state). I think many extension of the modal's functionality will depend on knowing its state.

If we stick to the prop, listening to the prop is inevitable, since the component probably won't unmount even invisible (v-show on the component itself probably does the same thing). It's paramount to remove the listener when note visible, since the listener is on document level, and may interfere/hijack other listener.

zackkrida commented:
@dos077 The purpose of this modal is to reduce the # of lines of code developers write and to make sure styling is consistent. In this case that means:

- our modal has the same overlay
- our modal is a white card with an 'x' in the right corner to dismiss

Having to style the content and close button each time we use the modal feels like a waste of effort to me. All of the ui library examples you gave would be fine as the _internal api_ for `AppModal`, and are actually very similar to what I have been asking for:

For example, this is the RMWC dialog example:

```jsx
function Example() {
  const [open, setOpen] = React.useState(false);
  return (
    <>
      <Dialog
        open={open}
        onClose={evt => {
          console.log(evt.detail.action);
          setOpen(false);
        }}
        onClosed={evt => console.log(evt.detail.action)}
      >
        <DialogTitle>Dialog Title</DialogTitle>
        <DialogContent>This is a standard dialog.</DialogContent>
        <DialogActions>
          <DialogButton action="close">Cancel</DialogButton>
          <DialogButton action="accept" isDefaultAction>
            Sweet!
          </DialogButton>
        </DialogActions>
      </Dialog>

      <Button raised onClick={() => setOpen(true)}>
        Open standard Dialog
      </Button>
    </>
  );
}
```

Now pretend the `<Dialog />` was made into its own component (we'll call it `<AppDialog />`):

```jsx
function Example() {
  const [open, setOpen] = React.useState(false);
  return (
    <>
      <AppDialog open={open} onClose={evt => setOpen(false)}}>
		This is a standard dialog.
	  </AppDialog>
		

      <Button raised onClick={() => setOpen(true)}>
        Open standard Dialog
      </Button>
    </>
  );
}
```


This is almost exactly the api I have proposed and requested for `AppModal` in vue:

```vue
<template>
  <app-modal id="meta-search-modal" v-if="open" @close="open = false">
	This is an app modal!
  </app-modal>

  <button @click="open = true">
    open app modal
  </button>
</template>
```

If you want to make additional style-agnostic primitive components with which to build `AppModal`, that is fine (<dialog-overlay>, <dialog-content>, etc.). The width can also be set by the children in the template, rather than in the modal component. But I would still like `AppModal` to:

- use `v-show` and `@close` for toggling visibility
  - close button lives within `AppModal`
- styles the modal contents as a card

As long as those requirements are met, you can implement however you like. Thank you!

dos077 commented:
The react component actually has all the changes that I recommended:
1. no id
2. render with a prop { open }
3. and has no fix close button/icon

These are all good patterns that allows flexibility and minimizing bugs. The absolute position button is not going to render correctly in all situation and in fact already has bugs with just the meta card.

![buttonoff](https://user-images.githubusercontent.com/45016741/88727660-086f3d80-d0fe-11ea-8a35-35aef6a59917.jpg)

The close button in FilterList component is another can of worm, as it switches styling from icon to  label depending on the breakpoint. To fit the current FilterList design in the modal card the close button will be split between two components, which will cause headaches for future maintenance.

Looking at other front end frame works, I guess what you are trying to achieve is [Bootstrap's modal](https://bootstrap-vue.org/docs/components/modal). In that case, the modal will need to add a title bar on top of the template content for proper spacing.

zackkrida commented:
I don't think continuing this discussion is productive. Please implement the modal how you like, and I will then create my desired `<app-modal>` as another layer of abstraction on top of your component(s). 

App modal is meant to be a _specific implementation of a modal with the content in a card and a close button in the top right_. It seems that you are only interested in building a more generic modal that can accommodate use cases we do not have yet.

The apis you are talking about are good and i agree with them (i also don't care about `id`s at all), but i want the app modal to be a very simple to use abstraction.

dos077 commented:
Did you look the [Bootstrap modal](https://bootstrap-vue.org/docs/components/modal) example I linked? If you want a card modal with a close button, you need to enforce the spacing with a title bar just like it. The title bar can be flat and colorless but it needs to float next to the close button so they won't overlap. An absolute overlay is going to cause problem since the template content won't know to space around it.

dos077 commented:
Well, whatever the current implementation is, it's not bug proof. Take a look at the screenshot, at about 600px(not an unreasonable res) wide the button overlaps with the title. I think I'll just implement an optional title bar trigger by a title prop. Having to try to padding the space correctly against an absolute overlay will be a re-usability nightmare.

![image](https://user-images.githubusercontent.com/45016741/88741453-b504eb80-d10d-11ea-876a-5be79d35a81c.png)

zackkrida commented:
Yes the current implementation is not bug proof. [This](https://codepen.io/zackkridacc/pen/bGEyQXN) is all I meant by padding, but again, please just take whatever implementation you like 😸. Some discussion is definitely healthy but with the number of issues and tickets we have I can't always go into this much depth in discussing an implementation.

dos077 commented:
- updated the modal to have a white bg
- adjusted the max dimension a little bit so the filter list don't clip in phones
- added a title prop
- added a optional titleBar and close button trigger by title prop
- added a header slot

Some how more bugs are pulled down while this is going on. DO NOT make the base filterList scroll-y. Scroll-y should be on the container of the list, which is the SearchGridFilter. The base list should just be a list that's as long as it needs.

- remove <div class="scroll-y"> at filterList. Scroll-y is added for all times on the parent component that's responsible for sizing

I am only trying to squash bugs here. I don't want to squash a dozen bugs in one pull request, but I also can't just submit buggy features.
